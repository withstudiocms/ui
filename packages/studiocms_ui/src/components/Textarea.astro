---
import type { HTMLAttributes } from 'astro/types';
import { generateID } from '../utils/generateID';

/**
 * Props for the textarea component
 */
interface Props extends HTMLAttributes<'textarea'> {
  /**
   * The label of the textarea.
   */
	label?: string;
  /**
   * The placeholder of the textarea.
   */
	placeholder?: string;
  /**
   * Whether the textarea is required. Defaults to `false`.
   */
	isRequired?: boolean;
  /**
   * Whether the textarea should take up the full width of its container. Defaults to `false`.
   */
	fullWidth?: boolean;
  /**
   * Whether the textarea should take up the full height of its container. Defaults to `false`.
   */
	fullHeight?: boolean;
  /**
   * Whether the textarea should be resizable. Defaults to `false`.
   */
	resize?: boolean;
  /**
   * The name attribute for the textarea. Useful for form submission.
   */
	name?: string;
  /**
   * Whether the textarea is disabled. Defaults to `false`.
   */
	disabled?: boolean;
  /**
   * The default value of the textarea.
   */
	defaultValue?: string;
}

const {
	label,
	placeholder,
	isRequired,
	fullWidth,
	fullHeight,
	resize,
	name = generateID('textarea'),
	disabled,
	defaultValue,
	...props
} = Astro.props;
---
<label 
  for={name} 
  class="sui-textarea-label"
  class:list={[
    fullWidth && "full-width",
    fullHeight && "full-height",
    resize && "resize",
    disabled && "disabled",
  ]}
>
  <span class="label">
    {label} <span class="req-star">{isRequired && "*"}</span>
  </span>
  <textarea
    placeholder={placeholder}
    name={name}
    id={name}
    class="sui-textarea"
    required={isRequired}
    disabled={disabled}
    {...props}
    
  >{defaultValue}</textarea>
</label>
<style>
  .sui-textarea-label {
    display: flex;
    flex-direction: column;
    gap: .25rem;
    max-width: 80ch;
  }

  .sui-textarea-label.disabled {
    opacity: 0.5;
    pointer-events: none;
    color: hsl(var(--text-muted));
  }

  .sui-textarea-label.full-width {
    width: 100%;
    max-width: none;
  }

  .sui-textarea-label.full-height {
    height: 100%;
  }

  .label {
    font-size: 14px;
  }

  .sui-textarea {
    padding: .65rem;
    border-radius: 8px;
    border: 1px solid hsl(var(--border));
    background: hsl(var(--background-step-2));
    color: hsl(var(--text-normal));
    transition: all .15s ease;
    resize: none;
    min-height: 12ch;
    width: 100%;
    height: 100%;
  }

  .sui-textarea:hover {
    background: hsl(var(--background-step-3));
  }

  .resize .sui-textarea {
    resize: both;
  }

  .sui-textarea:active,
  .sui-textarea:focus {
    border: 1px solid hsl(var(--primary-base));
    outline: none;
    background: hsl(var(--background-step-2));
  }

  .req-star {
    color: hsl(var(--danger-base));
    font-weight: 700;
  }
</style>
