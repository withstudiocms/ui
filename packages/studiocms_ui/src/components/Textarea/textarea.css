.sui-textarea-label {
  display: flex;
  flex-direction: column;
  gap: .25rem;
  max-width: 80ch;
}

.sui-textarea-label.disabled {
  opacity: 0.5;
  pointer-events: none;
  color: hsl(var(--text-muted));
}

.sui-textarea-label.full-width {
  width: 100%;
  max-width: none;
}

.sui-textarea-label.full-height {
  height: 100%;
}

.label {
  font-size: 14px;
}

.sui-textarea {
  padding: .65rem;
  border-radius: var(--radius-md);
  border: 1px solid hsl(var(--border));
  background: hsl(var(--background-step-2));
  color: hsl(var(--text-normal));
  transition: all .15s ease;
  resize: none;
  min-height: 12ch;
  width: 100%;
  height: 100%;
}

.sui-textarea:hover {
  background: hsl(var(--background-step-3));
}

.resize .sui-textarea {
  resize: both;
}

.sui-textarea:active,
.sui-textarea:focus {
  border: 1px solid hsl(var(--primary-base));
  outline: none;
  background: hsl(var(--background-step-2));
}

.req-star {
  color: hsl(var(--danger-base));
  font-weight: 700;
}
