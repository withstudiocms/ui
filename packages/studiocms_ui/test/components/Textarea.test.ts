import { describe, expect } from 'vitest';
import Textarea from '../../src/components/Textarea/Textarea.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const mockProps = {
	name: 'textarea',
};

describe('Textarea Component', () => {
	test('renders Textarea component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Textarea, 'Textarea', { props: mockProps });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ ...mockProps, label: 'Textarea Label' },
		{ ...mockProps, placeholder: 'Enter text here...' },
		{ ...mockProps, isRequired: true, label: 'Required Textarea' },
		{ ...mockProps, disabled: true, label: 'Disabled Textarea' },
	])('renders Textarea with different props "%s"', async (props, { renderComponent }) => {
		const result = await renderComponent(Textarea, 'Textarea', { props });
		expect(result).toMatchSnapshot();
	});
});
