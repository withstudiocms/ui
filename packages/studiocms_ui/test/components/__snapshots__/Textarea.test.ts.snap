// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea Component > renders Textarea component correctly 1`] = `"<label for="textarea" class="sui-textarea-label"> <span class="label">  <span class="req-star"></span> </span> <textarea name="textarea" id="textarea" class="sui-textarea"></textarea> </label>"`;

exports[`Textarea Component > renders Textarea with different props "{ name: 'textarea', disabled: true, label: 'Disabled Textarea' }" 1`] = `"<label for="textarea" class="sui-textarea-label disabled"> <span class="label"> Disabled Textarea <span class="req-star"></span> </span> <textarea name="textarea" id="textarea" class="sui-textarea" disabled></textarea> </label>"`;

exports[`Textarea Component > renders Textarea with different props "{ name: 'textarea', isRequired: true, label: 'Required Textarea' }" 1`] = `"<label for="textarea" class="sui-textarea-label"> <span class="label"> Required Textarea <span class="req-star">*</span> </span> <textarea name="textarea" id="textarea" class="sui-textarea" required></textarea> </label>"`;

exports[`Textarea Component > renders Textarea with different props "{ name: 'textarea', label: 'Textarea Label' }" 1`] = `"<label for="textarea" class="sui-textarea-label"> <span class="label"> Textarea Label <span class="req-star"></span> </span> <textarea name="textarea" id="textarea" class="sui-textarea"></textarea> </label>"`;

exports[`Textarea Component > renders Textarea with different props "{ name: 'textarea', placeholder: 'Enter text here...' }" 1`] = `"<label for="textarea" class="sui-textarea-label"> <span class="label">  <span class="req-star"></span> </span> <textarea placeholder="Enter text here..." name="textarea" id="textarea" class="sui-textarea"></textarea> </label>"`;
